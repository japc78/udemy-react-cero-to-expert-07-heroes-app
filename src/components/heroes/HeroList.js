import React, { useMemo } from 'react'
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher'
import { HeroCard } from './HeroCard';

export const HeroList = ({ publisher }) => {

	// Memorizamos el resultado siempre y cuando las dependencias se mantenga igual.
	const heroes = useMemo(() => getHeroesByPublisher( publisher ), [ publisher ]);
	// const heroes = getHeroesByPublisher( publisher);

	return (
		<div className='card-columns animate__animated animate__fadeIn'>
			{
				heroes.map( hero => <HeroCard key= {hero.id} { ...hero }/>)
			}
		</div>
	)
}
