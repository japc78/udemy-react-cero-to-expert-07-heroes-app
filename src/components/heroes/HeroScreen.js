import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router-dom'
import { getHeroesById } from '../../selectors/getHeroesById';

export const HeroScreen = ( { history }) => {

	// const params = useParams();
	// console.log(params);
	const { heroId } = useParams();


	// Memorizamos el resultado siempre y cuando las dependencias se mantenga igual.
	const hero = useMemo( () => getHeroesById( heroId ), [ heroId ]);
	// const hero = getHeroesById(heroId);

	console.log(hero);

	if (!hero) {
		return <Redirect to="/" />
	}

	const {
		superhero,
		publisher,
		alter_ego,
		first_appearance,
		characters
	} = hero;

	const handleReturn = () => {

		if( history.length <= 2) {
			history.push('/')
		} else {
			history.goBack();
		}
	}

	return (
		<div className="container">
			<div className="row mt-5">
				<div className="col-4">
					<img className="img-thumbnail animate__animated animate__fadeIn" src={`../assets/heroes/${heroId}.jpg`} alt={superhero}/>
				</div>


				<div className="col-8 animate__animated animate__fadeIn">
					<h3>{superhero}</h3>
					<ul className="list-group list-group-flush">
						<li className="list-group-item"><b>Alter ego:</b> {alter_ego}</li>
						<li className="list-group-item"><b>Publisher:</b> {publisher}</li>
						<li className="list-group-item"><b>First Appearance:</b> {first_appearance}</li>
					</ul>

					<h5>Characters</h5>
					<p>{characters}</p>
					<button
						className="btn btn-outline-info btn-sm"
						onClick={ handleReturn }
					>
						Volver
					</button>
				</div>
			</div>
		</div>
	)
}
