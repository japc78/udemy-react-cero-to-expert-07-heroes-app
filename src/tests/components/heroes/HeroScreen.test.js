import React from 'react'
import { mount, ReactWrapper } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import '@testing-library/jest-dom';
import { HeroScreen } from '../../../components/heroes/HeroScreen';


describe('Tests on <HeroScreen />', () => {


    let historyMock = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    }

    test('should shown the redirect component if there are no arguments in the Url', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries = {['/hero']}>
                <HeroScreen history = { historyMock }/>
            </MemoryRouter>
        );

        //Dos manera de hacerlo, que pinte el componete vacio.
        expect( wrapper ).toMatchSnapshot();

        // Que muestre el componente Redirect
        expect( wrapper.find('Redirect').exists()).toBe(true);
    })

    test('should shown a hero if the params exist', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries = {['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component = { HeroScreen }/>
            </MemoryRouter>
        );

        expect( wrapper.find('h3').exists()).toBe(true);
    })

    test('should return to before screen', () => {
        historyMock.length = 1;

        const wrapper = mount(
            <MemoryRouter initialEntries = {['/hero/marvel-spider']}>
                <Route
                    path="/hero/:heroId"
                    component = { (props) => <HeroScreen history = { historyMock }/> }
                />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();

        expect( historyMock.push ).toHaveBeenCalledWith('/');
        expect( historyMock.goBack ).not.toHaveBeenCalledWith();
    })

    test('should return to before screen called goBack without arguments', () => {
        historyMock.length = 10;

        const wrapper = mount(
            <MemoryRouter initialEntries = {['/hero/marvel-spider']}>
                <Route
                    path="/hero/:heroId"
                    component = { (props) => <HeroScreen history = { historyMock }/> }
                />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();

        expect( historyMock.push ).not.toHaveBeenCalled();
        expect( historyMock.goBack ).toHaveBeenCalled();
    })

    test('should called the redirect if hero not exist', () => {
        historyMock.length = 10;

        const wrapper = mount(
            <MemoryRouter initialEntries = {['/hero/hero-que-no-existe']}>
                <Route
                    path="/hero/:heroId"
                    component = { (props) => <HeroScreen history = { historyMock }/> }
                />
            </MemoryRouter>
        );

        // console.log(wrapper.html());
        expect( wrapper.text()).toBe('');
    })
});