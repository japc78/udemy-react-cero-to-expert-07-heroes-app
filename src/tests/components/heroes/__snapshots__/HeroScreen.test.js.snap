// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <HeroScreen /> should shown the redirect component if there are no arguments in the Url 1`] = `""`;
